### PR TITLE
[Higgs-Audio] bosonai/higgs-audio-v2-generation-3B-base TTS model support 

### PR DESCRIPTION
**Pull request by yuekaizhang**
The PR supported [higgs-audio](https://github.com/boson-ai/higgs-audio) to vllm-omni as a two-stage TTS pipeline (Stage 0 DualFFN talker + Stage 1 HiggsAudio codec).

#### Two modes implemented

- **Plain text → 24 kHz speech** :  [the_quick_brown_fox_jumps_over_the_lazy_dog.wav](https://github.com/user-attachments/files/28099174/the_quick_brown_fox_jumps_over_the_lazy_dog.wav)
- **Voice clone (ref_audio + ref_text)** : 
[00001_准确点说小森林是一部美食类电影食物佳肴贯穿了柿子的寒暑交替四十三餐.wav](https://github.com/user-attachments/files/28099188/00001_.wav)

#### Performance — 26 voice-clone prompts, 1× H100 80Gb

| Engine     | Batch | Wall (s) | Audio total (s) | Throughput (samp/s) |
|------------|------:|---------:|----------------:|--------------------:|
| vllm-omni  |     1 |  111.797 |         172.160 |               0.233 |
| vllm-omni  |     4 |   42.474 |         163.600 |             0.612 |
| vllm-omni  |     8 |   28.017 |         167.360 |           0.928 |

#### Optimizations status
- [x] Omnivoice verified
- [x] batch > 1 inference
- [ ] cuda graph (Not in this PR)
- [ ] async chunk (Not in this PR)

### CODE REVIEW

**Comment by linyueqian** [comment]
thats nice! @PopSoda2002 could you help review it as well?

File: vllm_omni/model_executor/models/omnivoice/omnivoice_decoder.py
Comment (linyueqian) [inline vllm_omni/model_executor/models/omnivoice/omnivoice_decoder.py:1]: why this is changed?

**Review by linyueqian — COMMENTED** [review]
## Two ship-blockers worth fixing before merge

**[important] 1. CI registration missing for the new e2e tests.**

The PR adds `tests/e2e/offline_inference/test_higgs_audio_v2.py` and `tests/e2e/online_serving/test_higgs_audio_v2.py`, both marked with `@pytest.mark.advanced_model` and `@hardware_test(res={"cuda": "H100"})`. Those markers do not, on their own, trigger a Buildkite step; the merge yaml needs an explicit pytest invocation, the same way peer TTS models are wired today:

```
.buildkite/test-merge.yml:173
  pytest -s -v tests/e2e/online_serving/test_qwen3_tts_customvoice.py tests/e2e/offline_inference/test_qwen3_tts_customvoice.py -m 'advanced_model and cuda' --run-level 'advanced_model'

.buildkite/test-merge.yml:335
  pytest -s -v tests/e2e/offline_inference/test_voxcpm2.py -m 'advanced_model and cuda' --run-level 'advanced_model'
```

`git diff origin/main..HEAD -- .buildkite` on this branch shows only the `v0.20.0 -> v0.21.0` bump and an XPU sampler toggle, with no higgs entry, so the new tests will never run on merge.

Suggested addition to `.buildkite/test-merge.yml` (following the qwen3_tts pattern):

```
pytest -s -v tests/e2e/online_serving/test_higgs_audio_v2.py tests/e2e/offline_inference/test_higgs_audio_v2.py -m 'advanced_model and cuda' --run-level 'advanced_model'
```

Without this, the validator-rejection guards and the voice-clone happy path go un-exercised on merge, so a future change that loosens `_validate_higgs_audio_v2_request` or breaks the prompt builder ships green.

**[important] 2. `pre-commit` is currently red.**

The latest `pre-commit` run reports `11 files reformatted, 1185 files left unchanged` and one unused-import removal (`to_dict` in `vllm_omni/model_executor/stage_input_processors/higgs_audio_v2.py`). It is a pure auto-format diff (ruff collapsing multi-line `raise ValueError(...)` calls into single lines plus the dead import), no logic change. A local `pre-commit run --all-files` followed by a commit of the staged changes should clear it.

File: vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py [hunk 1]
```diff
@@ -0,0 +1,1778 @@
...
+        valid = probs.sum(dim=-1) > 0
+        sampled = torch.empty(x.shape[0], dtype=torch.long, device=x.device)
+        if valid.any():
+            sampled[valid] = torch.multinomial(probs[valid], num_samples=1).squeeze(-1)
+        if (~valid).any():
+            sampled[~valid] = torch.argmax(logits[~valid].float(), dim=-1)
+        return sampled
+
+    @staticmethod
+    def _sample_audio_codes(
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py]: **[dead code]** `_sample_audio_codes` is unused. `sample()` only calls `_sample_audio_codes_upstream` (above, line 1344). This 41-line near-duplicate of the upstream version should be deleted, otherwise future edits will have to keep both copies in sync.

File: vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py [hunk 2]
```diff
@@ -0,0 +1,1778 @@
...
+            return "lm_head.weight"
+        # Dropped LlamaModel wrapper; text embedding + final norm now
+        # live directly on the talker (self.embed_tokens / self.norm).
+        if hf_name == "model.embed_tokens.weight":
+            return "embed_tokens.weight"
+        if hf_name == "model.norm.weight":
+            return "norm.weight"
+        # Per-codebook head, when the upstream exports it un-fused.
+        if hf_name == "codebook_head_0.weight":
+            return "audio_codebook0_head.weight"
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py]: **[dead mapping]** Returning `"audio_codebook0_head.weight"` here points at a model slot that does not exist after the R12 refactor that removed the codebook-0 split in favor of the fused `audio_lm_head` (see `_consume_fused_audio_head` and `__init__`). Since `_try_simple` checks `mapped in own_params` and fails silently, this branch is a no-op today and will hide a real loading bug the moment someone re-introduces a `codebook_head_0.weight` upstream. Remove this branch (or restore the slot if it is meant to come back).

File: vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py [hunk 3]
```diff
@@ -0,0 +1,1778 @@
...
+        - Audio-position logits at audio_token_id placeholders are also
+          read from ``lm_head``; the per-position codebook-0 routing path
+          lives in :meth:`audio_codebook0_logits` and is consumed by a
+          separate codebook-0 sampling adapter.
+
+        Call :meth:`audio_codebook0_logits` for the audio-side 1026-wide
+        head; that helper is exercised by unit tests + the upcoming
+        sampler dispatch.
+        """
+        _ = audio_token_mask  # The mask is consumed by audio_codebook0_logits, not here.
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py]: **[dead arg + broken xref]** `audio_token_mask` is declared then immediately discarded (`_ = audio_token_mask`). The docstring also tells readers to `Call :meth:audio_codebook0_logits` but the real method is `audio_codebook_logits` (no `0`). Drop the kwarg and fix the doc reference so callers don't think there is a mask-aware path here.

File: vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py [hunk 4]
```diff
@@ -0,0 +1,1778 @@
...
+          * Text MLP: HF ``mlp.gate_proj`` + ``mlp.up_proj`` ->
+            vLLM ``mlp.gate_up_proj`` (concatenated on dim 0).
+          * Audio MLP: HF ``audio_mlp.gate_proj`` + ``audio_mlp.up_proj`` ->
+            vLLM ``dual_ffns.<L>.audio_mlp.gate_up_proj`` (same shape transform).
+          * Layer norms: HF ``input_layernorm`` / ``post_attention_layernorm``
+            -> ``dual_ffns.<L>.{input,post_attention}_layernorm``. Audio norms
+            -> ``dual_ffns.<L>.{audio_input,audio_post_attention}_layernorm``.
+          * Audio token embedding: HF ``model.embed_audio_tokens.weight`` (with
+            optional outer ``embed_audio_tokens.`` prefix from upstream's
+            HiggsAudioV2Embeddings) -> ``self.embed_audio_tokens.weight``.
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py]: **[misleading doc]** The docstring routes `embed_audio_tokens` to `self.embed_audio_tokens.weight`, but no such attribute exists on the class (the `__init__` only declares `embed_tokens`, `audio_lm_head`, and `audio_codebook_embeddings`). The actual mapping in `_map_simple_name` already redirects everything to `audio_codebook_embeddings.weight`. Update the docstring to match.

File: vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py [hunk 5]
```diff
@@ -0,0 +1,1778 @@
...
+        # upstream samples its first audio frame from. Also reset any stale
+        # ``_audio_state`` and ``_slot_output_len`` for the slot, so a fresh
+        # request reusing a finished slot starts clean.
+        first_after_bos = getattr(self, "_last_first_audio_after_bos", None)
+        self._last_first_audio_after_bos = None
+        num_codebooks_local = int(self.config.num_codebooks)
+        bos_local = int(self.config.audio_stream_bos_id)
+        seeded_rows: list[int] = []
+        seeded_frame: torch.Tensor | None = None
+        # HF eager probe (tools/hf_first_audio_probe.py) confirmed that
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py]: **[dead ref]** The comment cites `tools/hf_first_audio_probe.py`, which is not in this PR (or in `main`). Either add the probe script to `tools/` or drop the reference; otherwise a reader chasing the citation hits a dead end.

File: tests/e2e/online_serving/test_higgs_audio_v2.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+
+v1 scope is plain text -> 24 kHz speech only. The model-aware request
+validator (vllm_omni/entrypoints/openai/serving_speech.py::_validate_higgs_audio_v2_request)
+rejects voice cloning fields, multi-speaker tags, language overrides, and
+task_type/voice selection with explicit 4xx — this suite exercises both the
+happy path (plain text in, audio bytes out) and the validator rejections,
+so a regression that loosens the schema will fail loudly.
+
+Unit-level coverage of the validator + tokenizer + DualFFN internals lives
+at tests/model_executor/models/higgs_audio_v2/test_higgs_audio_v2_units.py.
```
Comment (linyueqian) [inline tests/e2e/online_serving/test_higgs_audio_v2.py]: **[dead ref]** The module docstring promises unit coverage at `tests/model_executor/models/higgs_audio_v2/test_higgs_audio_v2_units.py`. That file is not in the PR. Either add the unit tests (DualFFN routing, fused-head reshape, delay-pattern revert, RAS, the boson `model.pth` to OmniVoice state-dict remap are all good candidates) or remove the pointer.

File: vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py [hunk 7]
```diff
@@ -0,0 +1,1778 @@
...
+        # HF seeds frame 0 with all-BOS at the audio_bos position and ignores
+        # the model's prediction there; the first real audio frame is sampled
+        # one step later at the audio_token_id position. R13's skip path is
+        # therefore the correct behavior. The HIGGS_R25_NO_SKIP env override
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/higgs_audio_v2/higgs_audio_v2_talker.py]: **[style]** Internal iteration labels (`R12`, `R13`, `HIGGS_R25_NO_SKIP`, `round-7`) appear at lines 298, 321, 695-697, 1165, 1244-1248. Per the repo convention they belong in the PR description and rot quickly in source. Suggested rewrite: keep the *why* sentences (they are useful), drop the round labels. Same pattern in the other call sites.

**Review by linyueqian — COMMENTED** [review]
Inline follow-ups for dead code and stale references found while running the model on H20.

**Review by linyueqian — COMMENTED** [review]
One more follow-up: integrate with the global uploaded-speaker cache so `voice=<name>` works after `POST /v1/audio/voices`.

File: vllm_omni/entrypoints/openai/serving_speech.py [hunk 8]
```diff
@@ -1383,6 +1394,160 @@ async def _build_moss_tts_params(self, request: OpenAICreateSpeechRequest) -> di
...
+        ``HiggsAudioV2TokenizerModel`` loaded from the k2-fsa/OmniVoice
+        ``audio_tokenizer/`` subdirectory), and attaches the encoded
+        ``audio_input_ids`` + ``audio_input_ids_mask`` tensors via
+        ``additional_information`` so the talker substitutes them at the
+        prompt-side audio placeholders.
+        """
+        from vllm_omni.model_executor.models.higgs_audio_v2.higgs_audio_v2_tokenizer import (
+            build_plain_text_prompt,
+            build_voice_clone_prompt,
+            input_ids_to_python_list,
```
Comment (linyueqian) [inline vllm_omni/entrypoints/openai/serving_speech.py:1521]: **[follow-up] Wire higgs into the global uploaded-speaker cache.**

Today this branch hard-rejects `request.voice`, so an uploaded speaker cannot be referenced by name (`POST /v1/audio/voices` -> later `voice=<name>` on `/v1/audio/speech`). To reach feature parity with qwen3_tts / voxcpm2:

1. Carve out `request.voice in self.uploaded_speakers` before this rejection (return `None` for that case so the request proceeds).
2. In `_build_higgs_audio_v2_params` (line 1526), when `request.voice` matches an uploaded entry and `request.ref_audio is None`, call `self._load_uploaded_audio(voice_lower)` (same pattern voxcpm2 uses at line 2261) to recover `(wav, sr)`, pull `ref_text` off the cached speaker payload, and pass them into `build_voice_clone_prompt`.
3. Confirm the upload endpoint persists `ref_text` on the speaker entry. voxcpm2 / voxtral already require it, but worth a quick check before the higgs branch reads it; otherwise persist it at upload time or run the model's ASR on the clip.

Not a blocker for the inline `ref_audio + ref_text` path that already works, just a follow-up so the cached-speaker flow behaves the same as the other TTS pipelines.

**Comment by linyueqian** [comment]
this pr is well written and look forward to your future contributions!

cc @hsliuustc0106 @Sy0307

**Comment by gcanlin** [comment]
Just curious, why does large batch size like `8` perform better than small one?

**Comment by yuekaizhang** [comment]
> Just curious, why does large batch size like `8` perform better than small one?

RTF here is for all requests: Total Generation Time / All Audios' Duration

**Comment by PopSoda2002** [comment]
Hi thanks for supporting this! LGTM if everything is passed! Remember to tell me if this is ready :)

File: vllm_omni/model_executor/models/omnivoice/omnivoice_decoder.py
Comment (yuekaizhang) [reply vllm_omni/model_executor/models/omnivoice/omnivoice_decoder.py:1]: Pure refactor, no behavior change. The RVQ codebook + DAC acoustic-decoder kernel that OmniVoice's decoder was carrying is identical to what the new higgs_audio_v2 Stage-1 needs, so I lifted the load/decode logic into `vllm_omni/model_executor/models/_shared/higgs_audio_decoder.py`

**Comment by yuekaizhang** [comment]
@linyueqian  Could you help review it again please? Thanks.

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
**BLOCKING:**

- **Gate Check** — docs/readthedocs.org is failing. Please fix the documentation build before requesting review.

- **Documentation** — The PR body is missing:
  1. Sample outputs (audio files for both plain text and voice cloning modes)
  2. Hardware specs for the RTF benchmark (e.g., "H100 80GB" not just "1x H100")
  3. e2e latency vs peak VRAM usage

- **Test Coverage** — The PR doesn't mention if online serving tests pass. Please verify that the new HiggsAudioV2 model works correctly with  endpoint.

- **Breaking Changes** — This PR modifies  and potentially affects existing OmniVoice models. Please verify backward compatibility.

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
BLOCKING: Docs build failing, missing PR body evidence, and omnivoice_decoder.py modification needs verification.

**Comment by yuekaizhang** [comment]
> BLOCKING: Docs build failing, missing PR body evidence, and omnivoice_decoder.py modification needs verification.

@hsliuustc0106  @linyueqian  Thanks. All resolved. Could you please help check it again?

**Comment by yuekaizhang** [comment]
> Hi thanks for supporting this! LGTM if everything is passed! Remember to tell me if this is ready :)

Hi @PopSoda2002 , I saw that Higgs Audio V3 will be released soon. I'm happy to support it if there are no immediate plans for vllm-omni.

**Review by linyueqian — APPROVED** [review]
LGTM, all prior review items addressed and CI is green across CUDA/AMD/Intel buildkites; thanks for the thorough cleanup pass.

**Review by hsliuustc0106 — APPROVED** [review]
lgtm
